modules/lang-painless/src/test/java/org/elasticsearch/painless/JsonTests.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+
+public class JsonTests extends ScriptTestCase {
+    public void testDump() {
+        // simple object dump
+        Object output = exec("Json.dump(params.data)", singletonMap("data", singletonMap("hello", "world")), true);
+        assertEquals("{\"hello\":\"world\"}", output);
+
+        output = exec("Json.dump(params.data)", singletonMap("data", singletonList(42)), true);
+        assertEquals("[42]", output);
+
+        // pretty print
+        output = exec("Json.dump(params.data, true)", singletonMap("data", singletonMap("hello", "world")), true);
+        assertEquals("{\n  \"hello\" : \"world\"\n}", output);
```

Comment:
Can you include a test with multiple keys on the same level to ensure that deterministic order (sorting) is done?

```json
{
  "key_a": "keys must be sorted",
  "key_b": 10,
  "key_c": 1
}
```

Reply:
@ypid-geberit thanks for the comment.

This change exposes internal elasticsearch json utility functions. Determinisitic ordering of output is not a requirement for this change.

Reply:
If you'd like to request a feature, please submit an issue so we can discuss and prioritize it.

Reply:
This requirement together with reasoning is part of #37585 that this PR intends to implement. So in this PR it should be stated if you see this as in scope of this PR or not.

Your comment suggests that you don’t. I see this as bad design. It will break your unit tests in hard to debug ways in the future.

Reply:
I'm not opposed to it, but it's not a requirement for this PR.  It can be added here or in a later PR.

Reply:
As I was working on this I notice there's optional sorting.  Makes sense to add a test for that.

Reply:
Let's take the ordering over in a new PR after we get this one in.  There's going to be a lot of SPI work here already.

We want two modes:
1) Preserve ordering.  `dumps(loads(String))` should be as close to a no-op as possible.
2) Force ordering. Probably need to add an additional flag.

Thanks.

Reply:
Json maps are, by design, unordered. Thus adding support for json serialization/deserialization to painless should be unencumbered by a requirement to support it, which is what this PR is about. Tests can be made robust to this, just as tests exist for maps in java which do not guarantee iteration order.

> Preserve ordering. dumps(loads(String)) should be as close to a no-op as possible.

I don't think this should be a requirement. That would mean the default is to preserve order, which is (1) outside the requirements of json and (2) brings the cost of a heavier weight data structure (eg linked hash map or tree map). Most users of json do not care about the order, since it is not guaranteed by any tools by default.

Reply:
> Tests can be made robust to this, just as tests exist for maps in java which do not guarantee iteration order.

That is a view that I can understand. Note that currently, it is not supported for testing watches this way however. If this issue next steps on my toes, I will consider changing the testing design to something where Watches would output valid JSON, and then compare the actual output with the expected output in Python. I will add it to https://github.com/elastic/examples/pull/239 (in case anyone cares 😉).

However, I am still very interested in a `sort_keys` (ref: https://docs.python.org/3/library/json.html#json.dump) feature. Feel free to ignore this requirement in this PR as discussed above.

> > Preserve ordering. dumps(loads(String)) should be as close to a no-op as possible.
>
> I don't think this should be a requirement.

Agreed.